File: app_test/examples/visual_test.rs
//! Generic visual test runner — launches N Slint windows sharing a reactive MockScribeHandle
//!
//! Usage:
//!   cargo run --example visual_test -p app_test -- <app_dir> [Name:role ...]
//!
//! Launched by scripts/visual_test.py which handles reload lifecycle.
//! No per-app Rust code needed. No recompilation for app changes.

use std::path::PathBuf;
use std::sync::{Arc, Mutex};

use lua_runtime::mock_scribe::{MockScribeHandle, MockScribeState};
use renderer_slint::{launch_test_slint_app, LaunchedApp};

struct Peer {
    name: String,
    role: String,
}

fn main() {
    tracing_subscriber::fmt::init();

    let args: Vec<String> = std::env::args().skip(1).collect();
    if args.is_empty() {
        eprintln!("Usage: visual_test <app_dir> [Name:role ...]");
        eprintln!();
        eprintln!("Examples:");
        eprintln!("  visual_test sample_apps/osvauld-demos/group-chat Alice:owner Bob:viewer");
        eprintln!("  visual_test sample_apps/my-shop/shop-owner ShopOwner:owner Customer:viewer");
        std::process::exit(1);
    }

    let app_dir = PathBuf::from(&args[0]);
    if !app_dir.join("manifest.json").exists() {
        eprintln!("Error: {} does not contain manifest.json", app_dir.display());
        std::process::exit(1);
    }

    let peers: Vec<Peer> = if args.len() > 1 {
        args[1..]
            .iter()
            .map(|arg| {
                let parts: Vec<&str> = arg.splitn(2, ':').collect();
                Peer {
                    name: parts[0].to_string(),
                    role: parts.get(1).unwrap_or(&"viewer").to_string(),
                }
            })
            .collect()
    } else {
        vec![
            Peer { name: "Alice".to_string(), role: "owner".to_string() },
            Peer { name: "Bob".to_string(), role: "viewer".to_string() },
        ]
    };

    let page_id = "test-page-1";
    let shared_state: Arc<Mutex<MockScribeState>> = MockScribeHandle::shared_state();
    shared_state.lock().unwrap().set_page_id(page_id);

    let mut apps: Vec<LaunchedApp> = Vec::new();

    for peer in &peers {
        let did = format!("did:key:{}", peer.name.to_lowercase());
        let scribe = MockScribeHandle::with_shared_state(shared_state.clone());

        eprintln!("Launching {} ({})", peer.name, peer.role);

        let app = launch_test_slint_app(
            &app_dir,
            page_id,
            &did,
            &peer.name,
            &peer.role,
            scribe,
        )
        .unwrap_or_else(|| {
            eprintln!("Failed to launch window for {}", peer.name);
            std::process::exit(1);
        });

        shared_state
            .lock()
            .unwrap()
            .register_subscriber(app.lua_tx.clone());

        apps.push(app);
    }

    eprintln!("{} window(s) launched. Close all windows to exit.", apps.len());

    slint::run_event_loop().unwrap();
}

File: scripts/visual_test.py
#!/usr/bin/env python3
"""Visual test launcher for Osvauld apps.

Launches the visual_test binary and manages reload lifecycle.
No Rust recompilation needed — change Lua/Slint and press 'r' to reload.

Usage:
    python scripts/visual_test.py <app_dir> [Name:role ...]
    python scripts/visual_test.py --preset chat
    python scripts/visual_test.py --preset ecomm
    python scripts/visual_test.py --preset chat10

Presets:
    chat    - group-chat with Alice (owner) + Bob (viewer)
    chat10  - group-chat with 10 users
    ecomm   - my-shop with ShopOwner (owner) + Customer (viewer)
    landing - sthalam landing page (read-only)
    docs    - protocol docs browser (read-only)

Controls:
    r + Enter  — reload all windows (re-reads Lua/Slint from disk)
    q + Enter  — quit
    Ctrl+C     — quit
"""

import subprocess
import sys
import threading
import signal
from pathlib import Path

SCRIPT_DIR = Path(__file__).parent
PROJECT_ROOT = SCRIPT_DIR.parent

PRESETS = {
    "chat": {
        "app_dir": "sample_apps/osvauld-demos/group-chat",
        "peers": ["Alice:owner", "Bob:viewer"],
    },
    "chat10": {
        "app_dir": "sample_apps/osvauld-demos/group-chat",
        "peers": [
            "Alice:owner", "Bob:viewer", "Carol:viewer", "Dave:viewer",
            "Eve:viewer", "Frank:viewer", "Grace:viewer", "Heidi:viewer",
            "Ivan:viewer", "Judy:viewer",
        ],
    },
    "ecomm": {
        "app_dir": "sample_apps/my-shop/shop-owner",
        "peers": ["ShopOwner:owner", "Customer:viewer"],
    },
    "landing": {
        "app_dir": "sample_apps/osvauld-demos/sthalam-landing",
        "peers": ["Alice:owner"],
    },
    "docs": {
        "app_dir": "sample_apps/osvauld-demos/protocol-docs",
        "peers": ["Alice:owner"],
    },
}

BINARY = PROJECT_ROOT / "target" / "debug" / "examples" / "visual_test"


def build_binary():
    """Build the visual_test binary if needed."""
    print("Building visual_test...")
    result = subprocess.run(
        ["cargo", "build", "--example", "visual_test", "-p", "app_test"],
        cwd=PROJECT_ROOT,
    )
    if result.returncode != 0:
        print("Build failed!")
        sys.exit(1)


def launch(app_dir: str, peers: list[str]) -> subprocess.Popen:
    """Launch the visual_test binary."""
    cmd = [str(BINARY), app_dir, *peers]
    return subprocess.Popen(cmd, cwd=PROJECT_ROOT)


def run(app_dir: str, peers: list[str]):
    """Main loop — launch binary, wait for input, reload on 'r'."""
    build_binary()

    proc = launch(app_dir, peers)
    print(f"\n  r + Enter = reload  |  q + Enter = quit  |  Ctrl+C = quit\n")

    def input_loop():
        """Read commands from stdin in a background thread."""
        nonlocal proc
        try:
            while True:
                line = input().strip().lower()
                if line == "r":
                    print("Reloading...")
                    proc.terminate()
                    proc.wait()
                    proc = launch(app_dir, peers)
                    print("Reloaded. (r = reload, q = quit)")
                elif line == "q":
                    proc.terminate()
                    proc.wait()
                    sys.exit(0)
        except (EOFError, KeyboardInterrupt):
            proc.terminate()

    thread = threading.Thread(target=input_loop, daemon=True)
    thread.start()

    # Wait for process to exit (user closed all windows)
    # Re-check after wait() returns — reload thread may have replaced proc
    import time
    while True:
        try:
            proc.wait()
            # Brief pause to let reload thread assign new proc
            time.sleep(0.3)
            # If reload thread spawned a new process, keep waiting
            if proc.poll() is None:
                continue
            break
        except KeyboardInterrupt:
            proc.terminate()
            proc.wait()
            break


def main():
    signal.signal(signal.SIGINT, lambda *_: sys.exit(0))

    if len(sys.argv) < 2:
        print(__doc__)
        sys.exit(1)

    if sys.argv[1] == "--preset":
        if len(sys.argv) < 3 or sys.argv[2] not in PRESETS:
            print(f"Available presets: {', '.join(PRESETS.keys())}")
            sys.exit(1)
        preset = PRESETS[sys.argv[2]]
        run(preset["app_dir"], preset["peers"])
    else:
        app_dir = sys.argv[1]
        peers = sys.argv[2:] if len(sys.argv) > 2 else ["Alice:owner", "Bob:viewer"]
        run(app_dir, peers)


if __name__ == "__main__":
    main()
